V1

lakenyon opened the pull request:
Initial 1.0 version.

lakenyon committed:
Project reorganization, transform from feature/initial branch.

lakenyon committed:
Minor renaming and folder org changes.

lakenyon committed:
Move various common structs from TPDataFood into their own files, and start adding validation, init's, etc.
Change all Float usage to Double.
Start removing id from TPData object init calls.

lakenyon committed:
Remove id from init of Basal, Food.
Add a food init that takes a nutrition struct.
Change more component structs to RawRepresentable, with validations: Nutrition, Energy, Fat, Protein. Add test of upload of more complete food item.

lakenyon committed:
Turn payload into a TPDataPayload struct with validation.
Restructure data structs/classes using TPData protocol and TPSampleData base class, to enable factoring out common validation/decoding/coding code from tidepool data types.
Continue creation of food common sub-type structs as TPData types with validation: Amount, Ingredient.
Use utility method of TPDataType enum to create debug descriptions of data types dynamically.

lakenyon committed:
Finish up TPDataFood sample type: ingredents array upload/download/description, various validations, etc.

lakenyon committed:
Break out TPUserProfile and TPUserSettings from TPUser. Convert over to using RawRepresentable like the user data objects.
TidepoolKitUI only does the login part now, skipping fetch of profile and settings info.
Fetch of TPUserProfile can use an optional passed in user id to allow fetching profiles for other access users.

darinkrauss committed:
Update project to generally conform with other Loop projects

- Remove deprecated UIApplication.shared.isNetworkActivityIndicatorVisible usage
- Update shared schemes for consistency
- Rename Example.app to TidepoolKitExample.app to avoid confusion
- Add required placeholder Default-568@2x.png
- Sort all groups by type
- Update project settings to match other Loop projects
- Add LICENSE file
- Add .gitignore

darinkrauss committed:
Minimum changes for initial integration into TidepoolService

lakenyon committed:
TPUserData -> TPDeviceData renaming.

lakenyon committed:
Added TPDeleteItem and TPDeleteItemArray, subclasses of TPDeviceData and TPDeviceDataArray respectively, and use these for item deletion: TIdepoolKit.deleteUserData now requires a TPDeleteItemArray rather than a TPDeviceDataArray. TPDeleteItemArray has a convenience init for transforming a TPDeviceDataArray into itself (for the case where there may be existing items that need to be deleted). Added a delete test where a TPDeleteItemArray is created from some TPDeleteItem objects with random id's.

lakenyon committed:
Simplify public api for upload/download, turning TPDeviceDataArray into a private APIDeviceDataArray, and using simple arrays of TPDeviceData for upload/download.

lakenyon committed:
Integration feedback: need to declare a logging interface and allow a client to specify a delegate to log to. If none specified, then don't log.
Defined a simple logging protocol TPKitLogging; optionally pass an object implementing this protocol to TidepoolKit and/or TidepoolKitUI at init time for logging; logger may also be optionally set/cleared later.
TPKitLoggerExample and TPKitLoggerUIExample are example loggers.
TidepoolKit.init() must now be called explicitly; after that TIdepoolKit.sharedInstance will return the singleton.
Start using standard header for new files.

lakenyon committed:
Replace TPDataType.getTypeFromDict with clearer getSelfFromDict, analog to addSelfToDict.

lakenyon committed:
Change API dir to Types dir (to avoid confusion, since networking files have API prefix).

lakenyon committed:
Keep field names and matching parameter names the same: carbs -> carbohydrate.

lakenyon committed:
Origin type also has a version string; origin type is an enum.

lakenyon committed:
Make TPDeleteItem a base class, rather than derived from TPDeviceData.

lakenyon committed:
Rework login to return a new TPSession object which may be persisted and used for subsequent logins. TidepoolKit no longer persists any data (this is left to the client). Login also takes a TidepoolServer enum to select a server (staging, production, etc.); TPKitExampleViewController shows how this may be persisted and used.
TPDataset has been redone as a RawRepresentable, and new TPDeduplicator and TPDatasetClient structs defined and are now members of a TPDataset. All are optional with defaults.
TidepoolKit.getDataset replaces currentUploadId, resetUploadId (no longer needed because TidepoolKit does not retain this), and configureUploadId: it tries to fetch a matching existing dataset from the service, and creates a new one if none exists.
TidepoolKitExample and TidepoolKitTests have been updated for all the changes and may be used as examples.
TidepoolKitError .badRequest now contains the index of failing samples for the putUserData failure case.
Methods putUserData and deleteUserdata now take a TPDataset parameter.
Methods getUserData, getUserProfileInfo, getUserSettingsInfo, getAccessUsers, getDataset now take a TPUser parameter.
All debugDescription calls replaced by description, and CustomStringConvertible conformance.

lakenyon committed:
Documented all the public TidepoolKit.swift methods and vars.
Changed some public method signatures to match the way Apple does it (e.g., getUserData(_ user: TPUser...) to getData(for user: TPUser...).
Change 404 errors into .dataNotFound errors; for getSettings call, turn this into a nil object return.
Change .serviceError to return option status code if available.
Change .badLoginResponse to return optional description.
Added test for getDatasets.
TidepoolKit.shared instance now always succeeds, creating the object if needed (with no logging).
Updated more headings.

lakenyon committed:
Update the rest of the headers to standard Tidepool format.

lakenyon committed:
In case of 404 error on fetch of existing dataset, create a new one. (Not tested!).

lakenyon committed:
Add test of posting a food item, getting it back from the service, and checking that all fields except id compare.

lakenyon committed:
Change logOut to asynchronously send a logout to the service. Note that this will invalidate any saved tokens (e.g., a persisted TPSession).
Added non-public clearSession method for testing to simulate restarting TidepoolKit, in order to test logging in with a persisted session.

lakenyon committed:
Minor cleanups.

lakenyon committed:
Added refreshSession to check whether current session auth token is valid.

lakenyon committed:
Fix tests to pass. Consolidate and simplify some networking code. Change refreshSession to log out if service reports that the token is invalid.

lakenyon committed:
Get tests running on device (ensure framework binaries are embedding in app, signing is correct, etc) using TidepoolKitExample as host (exit early from view controller before host has initialized TidepoolKit).
On authorization errors, always clear session to avoid repeated service calls in case of an expired token.

lakenyon committed:
Factor out some json parsing into dictFromJsonData and  dictArrayFromJsonData methods.
Use try? in some places to reduce code complexity.

lakenyon committed:
Implement base struct for association and its array (still needs test).
Misc cleanup:
Break out TPPatient into its own file.
Move validations (still need to decide if these are needed) to a more appropriate class.
Make TPSession variables public; remove @testable from test code so lack of public is caught in test code.

lakenyon committed:
Remove TidepoolKit.sharedInstance method, no longer require it to be a singleton...

lakenyon committed:
Fix tests for sharedInstance change.

lakenyon committed:
Fix a couple bugs with rawValue of session and user.
Added persist of TPSession in TidepoolKitExample, with auth token stored separately.

lakenyon committed:
Fix bug Darin reported where initial create of dataset fails: don't put nil values in rawValue [String: Any] dictionaries (i.e., don't use 'as Any').

lakenyon committed:
Add support for base data assocications and locations (and its base structs). Untested.

lakenyon committed:
Allow client to pass in default server to TidepoolKitUI - no longer persist this.

lakenyon committed:
TidepoolKit refreshSession method now returns a TPSession (also available via the currentSession var), with a refreshed token and user email (now makes a second call to get auth/user).
Changes TPUser.userName to userEmail.
Fixed bug where code didn't return after checking for offline/logged out.
Started splitting out tests into more files to better organize around test areas. Clean up some of the test expectation.fulfill calls, placing them where async activity completes, not necessarily where test passes.

lakenyon committed:
Bolus types, work in progress.

lakenyon committed:
Finish bolus types, tests.

lakenyon committed:
Add support for basal types. Start adding tests.

lakenyon committed:
Finish basal types: add support for percent, suppressed, second level of suppressed, add more tests, etc.
Moving from validation on the client (other than basic types and enums) to service side validation, remove optional inits of device data objects as these will no longer fail (except the inits from rawValue), but leave LogError's in for now.
Add support for more base types: clockDriftOffset (leave an Int for now), conversionOffset, deviceId, deviceTime (leave a String for now), guid, timeZone (leave a String for now), and timeZoneOffset (leave an Int for now).
When comparing datasets for equivalence, ignore the dataset client version number.

lakenyon committed:
Add tests for Location types, and Associations, fix a couple bugs in both revealed in testing.
Add test support for injecting a mock reachability object, and use it to add tests for offline handling of log in and other network calls. Fix a couple bugs exposed.
Start adding support for mocking service with an alternate URLSession.
Change data upload to use a background URL session and task, pushing the data into a file.
Turn on code coverage for unit tests.

lakenyon committed:
Added test of uploading samples including an out-of-range value, verifying bad sample index array is correctly parsed.
Extended .badRequest error to include the whole response as well as the bad sample array since we are now relying on the service to do the value checking.

lakenyon committed:
Simplify network mocking interface, moving it from URLSession level to network call level.
Minor networking code simplifications.
Added optional specification of dispatchQueue to use for network completions to make threading more clear.
Added login tests using mock interface: this can be easily extended to other tests to allow testing w/o a network/service, and to cover error cases that would be difficult to test with a real network/service.

darinkrauss commented on Playgrounds/Codables.swift:
```diff
@@ -0,0 +1,125 @@
+import UIKit
```
Can we drop the `Playgrounds` directory and `TidepoolKit.playground`? No other Loop related repo has playgrounds committed.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Types/TPDeviceData/TPDataCbg.swift:
```diff
@@ -0,0 +1,89 @@
+    case millimolesPerLiter = "mmol/L"
+    
+    // service syntax check
+    func inRange(_ value: Double) -> Bool {
```
Let's drop the range validation.

lakenyon replied:
Done. Removed all validations on object inits except from raw values.

darinkrauss commented on TidepoolKit/Types/TPDeviceData/TPDataCbg.swift:
```diff
@@ -0,0 +1,89 @@
+    public init(time: Date, value: Double, units: TPCbgUnit) {
+        self.value = value
+        self.units = units
+        if !self.units.inRange(self.value) {
```
Drop the range validation.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Types/TPDeviceData/TPDataCbg.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    //
+    // MARK: - RawRepresentable
```
I think we can drop a lot of the `MARK`s. In general, we've only used those to separate out different chunks of logic in large files.

lakenyon replied:
I've now mostly removed these except for a couple large files where they do help show overall organization and help to get the reader to a specific area, and where they are used to set off protocol conformance (mainly for rawRepresentable, where that can't be done via an extension of a class). Also removed the extra // lines around the MARK statements to match usage in Loop.

darinkrauss commented on TidepoolKit/Types/TPDeviceData/TPDataCbg.swift:
```diff
@@ -0,0 +1,89 @@
+        self.value = value.doubleValue
+        self.units = units
+        
+        // base properties in superclass...
```
Not sure that these types of comments are necessary. There aren't a lot of inline comments in Loop. Pretty much just limited to areas of code where the logic is complicated or the code is unexpected, not where the code is expected.

lakenyon replied:
Removed a number of these, except where they help me remember what's going on ;-)

darinkrauss commented on TidepoolKit/Types/TPDeviceData/TPDataCbg.swift:
```diff
@@ -0,0 +1,89 @@
+    
+    override public var rawValue: RawValue {
+        // start with common data
+        var dict = super.rawValue
```
Generally, we've used `rawValue` local variable rather than `dict` in this situation.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Types/TPDeviceData/TPDataCbg.swift:
```diff
@@ -0,0 +1,89 @@
+import Foundation
+import HealthKit
+
+public enum TPCbgUnit: String, Encodable {
```
This isn't limited to `cbg`, but rather common to all blood glucose values (regardless of source, `cbg`, `bgm`, or otherwise).

lakenyon replied:
Moved to separate file.

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+    public let users: [TPUser]
+        
+    //
+    // MARK: - methods private to framework!
```
Probably not needed.

lakenyon replied:
Removed.

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+    //
+    
+    class func urlExtension(forUser userId: String) -> String {
+        let urlExtension = "/access/groups/" + userId
```
Consider `return "/access/groups/\(userId)"`.

lakenyon replied:
Why?

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+    // MARK: - methods private to framework!
+    //
+
+    init(_ users: [TPUser]) {
```
I would even go so far as to make it `private` since the only time it will be constructed is from the `fromJsonData` below.

lakenyon replied:
fromJsonData is now in an extension; this is not public, but needs to be at least "internal" for it to be accessible in fromJsonData.

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+    // MARK: - TPFetchable protocol conformance methods
+    //
+    
+    class func urlExtension(forUser userId: String) -> String {
```
Initially confused by the `urlExtension` name. Consider `urlPath` as a more accurate name.

lakenyon replied:
Renamed.

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+}
+
+
+/*
```
I would avoid all of the JSON examples. They are likely to just go stale over time and cause confusion. The API spec in Stoplight is the definitive API documentation.

lakenyon replied:
Removed.

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+            if let keyStr = key as? String {
+                let user = TPUser(keyStr)
+                users.append(user)
+                print("adding user: \(user)")
```
Probably should drop most of the `print` statements.

lakenyon replied:
Removed and/or changed to Log statements.

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+        }
+        var users: [TPUser] = []
+        for key in jsonDict.keyEnumerator() {
+            if let keyStr = key as? String {
```
It took me a while to figure this out, but `keyStr` is really `userId`, isn't it? If so, I'd user `userId` here to make it more clear.

lakenyon replied:
Yes, changed!

darinkrauss commented on TidepoolKit/Networking/APIAccessUsers.swift:
```diff
@@ -0,0 +1,68 @@
+        var users: [TPUser] = []
+        for key in jsonDict.keyEnumerator() {
+            if let keyStr = key as? String {
+                let user = TPUser(keyStr)
```
So we are only capturing the `userId` here? Not the rest of the structure?

lakenyon replied:
Yes, that is all that has been used in Tidepool Mobile to date. If Loop or other clients need more, we should prioritize.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+import Foundation
+
+typealias NetworkRequestCompletionHandler = (Data?, URLResponse?, Error?) -> Void
+protocol TidepoolNetworkInterface {
```
[nit] Blank line between `NetworkRequestCompletionHandler` and `TidepoolNetworkInterface`?

lakenyon replied:
Added.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+
+import Foundation
+
+typealias NetworkRequestCompletionHandler = (Data?, URLResponse?, Error?) -> Void
```
I would make a `struct` that joins the `Data` and `URLResponse` objects. Maybe call it `ResponseData`. Then change the handler to `Result<ResponseData, Error>. Then, ditch the `typealias` because it can be easily represented as `(Result<ResponseData, Error>) -> Void`.

This also makes it clear that there will either be a `ResponseData` with `Data` and `URLResponse` *OR* and `Error`.

lakenyon replied:
This is a common trio of values coming from URLSession calls, the code dispatches these at a very low level and quickly incorporates them into a struct along these lines. Not sure it's worth the effort to change this.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+    var apiQueue: DispatchQueue
+    
+    // Base URL for API calls, set during initialization
+    var baseUrlString: String? {
```
Consider making this an entirely computed property backed by the `baseURL` property?

lakenyon replied:
Now just a string, used in 2 places to create a URL.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+    var baseUrlString: String? {
+        didSet {
+            if let urlStr = baseUrlString {
+                self.baseUrl = URL(string: urlStr)!
```
What if this fails?

lakenyon replied:
Should only be due to a coding error - ! causing a crash should be fine. Anyway, now just a string since baseUrl wasn't being used anymore.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+            }
+        }
+    }
+    var baseUrl: URL?
```
Actually, why is there even a `baseUrl` in the first place? I don't see it being used.

lakenyon replied:
Removed, it was a remnant of a previous version...

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+    var networkRequestHandler: TidepoolNetworkInterface
+    
+    init(queue: DispatchQueue) {
+        LogInfo("")
```
???

lakenyon replied:
LogInfo takes a string; with a nil string, the function, file, etc are still logged.

darinkrauss replied:
Understood. Since it is not obvious (to me, at least), please replace occurrences of `LogInfo("")` with a reasonable message such as `LogInfo("Initializing APIConnector")` or something like that. It will help logging and well as code readability.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+    // Dictionary of servers and their base URLs
+    private func serverUrl(_ server: TidepoolServer) -> String {
+        switch server {
+        case .development: return "https://dev-api.tidepool.org"
```
[nit] Separate lines?

lakenyon replied:
Changed, but this is going away with the Kubernetes support.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+        case .production: return "https://api.tidepool.org"
+        }
+    }
+    private let kDefaultServer: TidepoolServer = .staging
```
I would add the default to the `TidepoolServer` enum itself as a computed property so it could be used like `TidepoolService.default`.

lakenyon replied:
Again, will be changing with the Kubernetes support.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,994 @@
+    }
+    
+    func clearSession() {
+        LogVerbose("")
```
???

lakenyon replied:
Same as above.

lakenyon replied:
Done.

darinkrauss committed:
Update project with minor setting changes for consistency

lakenyon committed:
Fix some text disappearing issues for iOS 13 dark mode.

lakenyon committed:
PR review feedback: try to make networking classes more uniform as extensions of TPxxx classes for TPUploadable, TPFetchable conformance.

lakenyon committed:
Remove all checking of TPData objects at init time, leaving this for service.

darinkrauss reviewed (changes requested):
Partial review.

lakenyon committed:
PR review feedback: removed playground. Removed excess MARK statements. Removed unused baseURL variable.

lakenyon committed:
PR review feedback: some MARK cleanup, consistency.

lakenyon committed:
PR request feedback: reduce use of comments where obvious. Get rid of NSLogs in framework, print in test code.

lakenyon committed:
PR review feedback: dict, resultDict -> rawValue.

lakenyon committed:
PR review feedback: Pulled out TPCbgUnit to a common enums file. Renamed protocol function urlExtension to urlPath. Removed some checking code that wasn't being used in TPDataEnergy. Changed keyStr to userId in TPAccessUsersExtension.

lakenyon committed:
Remove playground files.

lakenyon committed:
Use a standard approach for networking support for fetching/uploading arrays of objects; since extending Array can only work for one particular type of array, just use a simple class containing an array, which can easily be extended to support any protocols desired (though maybe not quite as elegant).

lakenyon committed:
Kubernetes support: switch from a fixed enum of (production, development, staging...) for servers, to a host string. Populate possible choices from a dynamic DSN lookup of SRV records at environments-srv.tidepool.org. Hardwire one default (currently staging, but should be changed on release to production!).

lakenyon committed:
Revert check-in of person team credentials needed to run on device.

darinkrauss commented on TidepoolKitUITests/TidepoolKitUITests.swift:
```diff
@@ -0,0 +1,34 @@
+//
```
Remove since there are no tests in this file?

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+import TidepoolKit
+
+
```
[nit] Remove extra blank line?

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.swift:
```diff
@@ -0,0 +1,34 @@
+
+public class TidepoolKitUI {
+    
+    private var tpKit: TidepoolKit
```
[nit] Rename `tpKit` -> `tidepoolKit` throughout.

lakenyon replied:
Changed this for TidepoolKit and TIdepoolKitUI files. Left it for test files, I find it more readable.

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.swift:
```diff
@@ -0,0 +1,34 @@
+    public func logInViewController(loginSignupDelegate: LoginSignupDelegate? = nil, defaultServerHost: String?) -> UIViewController {
+        let loginViewController = UIStoryboard(name: "LoginSignup", bundle: Bundle(for: LoginViewController.self)).instantiateInitialViewController() as! LoginViewController
+        loginViewController.loginSignupDelegate = loginSignupDelegate
+        loginViewController.tpKit = self.tpKit
```
[nit] Don't need `self.` here.

lakenyon replied:
True. I often do this so the reader doesn't need to check whether this is a local vs instance variable. Changed here.

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.swift:
```diff
@@ -0,0 +1,34 @@
+}
+
+// global logging protocol, optional...
+var clientLogger: TPKitLogging?
```
[nit] Its not really clear what `clientLogger` means in this context. Perhaps `globalLogger` might be more appropriate?

lakenyon replied:
Changed. (Meaning was this was a logger coming from the framework client).

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.swift:
```diff
@@ -0,0 +1,34 @@
+        self.tpKit = tpKit
+    }
+    
+    public func logInViewController(loginSignupDelegate: LoginSignupDelegate? = nil, defaultServerHost: String?) -> UIViewController {
```
[nit] Rename `defaultServerHost` -> `serverHost` since that is the name of the property on the `LoginViewController`.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.swift:
```diff
@@ -0,0 +1,34 @@
+//
+//  TidepoolKitUI.swift
+//  TidepoolKit
```
[nit] TidepoolKitUI?

darinkrauss commented on TidepoolKitUI/TidepoolKitUI.h:
```diff
@@ -0,0 +1,19 @@
+//
+//  TidepoolKitUI.h
+//  TidepoolKit
```
[nit] TidepoolKitUI?

lakenyon replied:
Hmmm... thought this was supposed to be the project name. Fixed.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+//
+//  LoginViewController.swift
+//  TidepoolKit
```
[nit] TidepoolKitUI?

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+import UIKit
+import TidepoolKit
+
+
```
[nit] Please remove all double blank lines with a single blank link throughout all files with a quick find/replace. I believe we decided that this extra spacing was not desired.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+
+class LoginViewController: UIViewController {
+
+    var loginSignupDelegate: LoginSignupDelegate?
```
I think this should be `weak`?

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+class LoginViewController: UIViewController {
+
+    var loginSignupDelegate: LoginSignupDelegate?
+    var tpKit: TidepoolKit!
```
[nit] `tidepoolKit`?

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+    }
+    
+    private var currentServerHost: String {
+        return self.serverHost ?? tpKit.currentServerHost
```
I don't think this should be reaching into `tidepoolKit` to get the current server host. The UI should be passed the server host to be used as the initial value. Then, this UI should maintain whatever is currently selected while the UI is visible. Finally, when the UI completes and calls the delegate it should pass back the final selected server host.

lakenyon replied:
It already passes back the selected server host in the session object. The problem is that there are two "default" server hosts: the last selected one (for debug login) and the production host, which is not returned by the dynamic DNS SRV record lookup. 

So I changed TidepoolKit to be ignorant of any of these, and to just require a server host to be passed into login. I also changed TidepoolKitUI to ensure that the production server host appears in the debug UI host selection list (test scenario where a non-production host has been selected previously, but now the user wants to go back to production, but the DNS lookup does not provide this). The client of TidepoolKitUI will now also have to conjure up a server host if it has not yet persisted one: this should be production for shipping code. 

I'm not overly fond of this solution because now the production server host needs to be hardwired in each client of TidepoolKitUI as well as in TidepoolKitUI itself (before this change it was just kept in TidepoolKit) - so, open to suggestions here...

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+        configureForReachability()
+        updateButtonStates()
+        let notificationCenter = NotificationCenter.default
+        notificationCenter.addObserver(self, selector: #selector(LoginViewController.textFieldDidChange), name: UITextField.textDidChangeNotification, object: nil)
```
Why are you using the notification center for the text field did change? Wouldn't it be more localized if this view controller were to implement the `UITextFieldDelegate` and handle it that way?

lakenyon replied:
Using notifications is an anachronism. Changed to just hook up the editing changed connection.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+        notificationCenter.addObserver(self, selector: #selector(LoginViewController.textFieldDidChange), name: UITextField.textDidChangeNotification, object: nil)
+        notificationCenter.addObserver(self, selector: #selector(LoginViewController.reachabilityChanged(_:)), name: ReachabilityChangedNotification, object: nil)
+        
+        self.configureCurrentServerButton()
```
No `self.` needed here. Plus, I'd put the button state/configuration code together (eg. up before `updateButtonStates`).

lakenyon replied:
Done.

darinkrauss commented on TidepoolKitUI/LoginViewController.swift:
```diff
@@ -0,0 +1,206 @@
+    static var firstTime = true
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        if LoginViewController.firstTime {
```
Not sure what the purpose is here. Could you explain?

lakenyon replied:
Removed, this was old code.

darinkrauss commented on TidepoolKitExample/AppDelegate.swift:
```diff
@@ -0,0 +1,46 @@
+//
+//  AppDelegate.swift
+//  TidepoolKit
```
[nit] TidepoolKitExample?

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+        }
+    }
+    
+    private let kSessionTokenHeaderId = "X-Tidepool-Session-Token"
```
The Loop project does not use the `k` prefix for constants. I suggest `tidepoolSessionTokenHeaderKey`. 

More importantly, we shouldn't have two different constants here either. We cannot assume the server will always respond with all lowercase (as HTTP header keys are case insensitive and the code needs to accommodate that). So, just define the canonical header (`private let tidepoolSessionTokenHeaderKey = "X-Tidepool-Session-Token") and then do lowercase manipulations, if necessary.

For example, the code below:
```
guard let token = httpResponse.allHeaderFields[self.kSessionTokenResponseId] as? String else
```
Could become something like:
```
guard let token = httpResponse.allHeaderFields.first(where: { return ($0 as? String)?.lowercased == tidepoolSessionTokenHeaderKey.lowercased })?.key as? String else
```

(I'm sure you can make that code cleaner, but you get the idea.)

lakenyon replied:
Ok, changed to just use the lower-cased version, and lower-cased the checks on the response coming back. Removed all k prefixing from constants.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+            
+            guard error == nil else {
+                let error = error!
+                LogError("Login post failed with error: \(error)!")
```
I don't think you need all of these `LogError` statements followed by a `completion(.failure...`). If the client received an error on completion then it can decide if it wants to log it or not. The other Loop libraries are not "chatty". It is up to the caller of the library whether to log anything or not.

lakenyon replied:
Removed some of these that won't affect ability to debug issues.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+            guard error == nil else {
+                let error = error!
+                LogError("Login post failed with error: \(error)!")
+                completion(Result.failure(error))
```
[nit] You can remove the `Result` here. So, just `completion(.failure(error))`. Same throughout the rest of the code.

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+    
+    private let kSessionTokenHeaderId = "X-Tidepool-Session-Token"
+    private let kSessionTokenResponseId = "x-tidepool-session-token"
+    let kDefaultServerHost = "stg-api.tidepool.org"
```
This needs to be remove. As far as I'm concerned, there should be no default server host. It should be a required parameter by the client.

lakenyon replied:
Changed, see above.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+}
+
+public protocol ReachabilitySource {
+    func serviceIsReachable() -> Bool
```
Why not just call this `isReachable`. Then `Reachability` already implements it and you don't need another function that does the same thing.

darinkrauss replied:
Specifically, just:
```
var isReachable: Bool
```

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+    public func configureNotifier(_ on: Bool) -> Bool {
+        if on {
+            do {
+                try self.startNotifier()
```
In general, the Loop projects doe not use `self.` prefix except under 3 conditions: 1) in the initializer when *setting* instance properties; 2) when *setting* instance properties where `self.` is required for clarify in other places than an initializer; 3) when in a closure that requires `self.` (and it should be defined as `weak`).

lakenyon replied:
I've learned to use it in more cases, but fine to elide it where not absolutely required (I do find it makes the code reading easier, especially when you aren't familiar with it).

lakenyon replied:
Done.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+class APIConnector {
+    
+    /// Current session: non-nil when "loggedIn", nil when "loggedOut"
+    var session: TPSession?
```
I thought the session was going to be provided by the client when calling any API so TidepoolKit was more-or-less stateless?

lakenyon replied:
Change in progress (fairly major, still adjusting all the test code)...

Btw, I thought we had just agreed that TidepoolKit would not persist any data itself. Even with this change there is state: the logger, any of the test protocol objects, any call in progress. The service itself has this state of a token being valid or not.

lakenyon replied:
Ok, now most calls require a TPSession parameter, and APIConnect no longer has this session state. Calls relating to "log-in" state have been removed, along with the notification of log-in state change.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+        }
+    }
+    
+    func login(with session: TPSession) -> Result<TPSession, TidepoolKitError> {
```
In general, why do some things have a `TP` prefix and others have `TidepoolKit` prefix?

lakenyon replied:
Generally TP prefix is what I've been using. I did use TidepoolKit as a prefix for the logging protocol and error type, changed both now to TP. Basically TP is the shortened version of TidepoolKit when the name gets long, but probably good to be consistent. So TidepoolKit will be used for the project names, and the two classes TidepoolKit and TidepoolKitUI; TP will be used for the public files associated with TidepoolKit.

darinkrauss commented on TidepoolKit/Networking/APIConnect.swift:
```diff
@@ -0,0 +1,950 @@
+    }
+    
+    func login(with session: TPSession) -> Result<TPSession, TidepoolKitError> {
+        if self.session != nil {
```
`guard session == nil else {` ?

lakenyon replied:
Yes, fixed.

darinkrauss reviewed (changes requested):
Some new comments so you can start responding and/or updating while I continue on with other comments.

lakenyon committed:
PR Review feedback: LogInfo(""):
LogVerbose("") changed to something more meaningful.
clientLogger -> globalLogger.
TPKit -> TidepoolKit (in TidepoolKit, TidepoolKitUI). TPKitLogging -> TPLogging.
Remove self. where unnecessary for compiler.
Use qa2.development.tidepool.org instead of stg-api.tidepool.org.

lakenyon committed:
PR Review feedback: LogInfo(""):
Update TidepoolKitUI headers to TidepoolKitUI from TidepoolKit.
defaultServerHost -> serverHost.
Change all double blank lines to single blank lines.

lakenyon committed:
PR Review feedback: LogInfo(""):
Make TidepoolKit require service host to be passed in at login time. TidepoolKit now inserts production server host into debug ui server host selection. Client of TidepoolKitUI is now required to pass either the production server host or last selected host into TidepoolKitUI.

lakenyon committed:
PR Review feedback: LogInfo(""):
Use editingChanged connection rather than notification in login UI for monitoring edit fields.
Use lower-cased token key, and lower-case responses in checking for returned token. Remove "k" prefix from constants.
Adjust all headers to have name of target rather than project (e.g., TidepoolKit -> TidepoolKitExample).
Remove some unused code from TidepoolKit LoginViewController.

lakenyon committed:
PR Review feedback: removed some of the more chatty LogError statements.

lakenyon committed:
PR Review feedback: slight simplification of ReachabilitySource protocol.

lakenyon committed:
PR Review Feedback: remove current session state in APIConnect. This includes removing TidepoolLoginChangedNotification (no longer tracks a "login" state), several calls/variables from the public interface (isLoggedIn, loggedInUser, currentSession, login:withSession, and the clearSession test function. Most TidepoolKit.swift public functions now require an explicit TPSession parameter to be passed (e.g., using a with: argument label). Updated all tests and the example application.

lakenyon commented:
Closing this initial PR, and creating a new one for everyone.
